### DISCUSSION TITLE
Does the source timestamp of a DataValue get overridden?

### DISCUSSION THREAD

**Discussion by prathik-quartic** [FreeOpcUa/opcua-asyncio#604] [Q&A] [answered]
I start a opcua server and create 20 nodes. I connect to the server as a client and continuously write data to all the 20 nodes at a update interval of 500 milliseconds. I am trying to make sure while writing data each time to all the 20 nodes, they have the same source timestamp. Here is the code snippet that tries to write data to all 20 tags.
```
async def write_data_to_each_node(node_id, value, source_timestamp):
    print(f"Source Timestamp: {source_timestamp.timestamp() * 1000}")
    datetime_current = datetime.utcnow()
    data_timestamp = int(datetime_current.timestamp() * 1000)
    print(f'writing {str(node_id)} with {value} at {data_timestamp}')

    data_value = ua.DataValue(value)
    data_value.SourceTimestamp = source_timestamp
    # data_value.ServerTimestamp = datetime_current
    node_id.set_value(value)

async def write_data(self, nodes):
    await asyncio.gather(*[write_data_to_each_node(node, values[i], datetime.utcnow()) for i, node in enumerate(nodes)])
```

When I fetch these data as a client, I observed that the source_timestamp that I assigned to the node was not the same as the source_timestamp of the data fetched. 
Is the source_timestamp of the DataValue overridden by the server? Or am I might be missing something?

**Comment by oroulet** [comment, marked as answer]
It should not be I have seen some merge request touching that lately, maybe there is a bug...

**Reply by prathik-quartic** [reply]
Oh okay. So how to proceed? Should I create an issue for someone to investigate?

**Reply by AndreasHeine** [reply]
propably here:
https://github.com/FreeOpcUa/opcua-asyncio/blob/7a86b031f2ae23463dccaaf2ca10a4b74a044e71/asyncua/common/node.py#L220
https://github.com/FreeOpcUa/opcua-asyncio/blob/7a86b031f2ae23463dccaaf2ca10a4b74a044e71/asyncua/common/ua_utils.py#L16

pass a DataValue as value to set_value and it should work ;)

**Reply by prathik-quartic** [reply]
@AndreasHeine In my code snippet that I mentioned, I am passing the DataValue to set_value.

**Reply by AndreasHeine** [reply]
in the snipet you make: node_id.set_value(value)
should be: node_id.set_value(data_value)

**Reply by prathik-quartic** [reply]
Ah damn. My bad. Not sure how I missed such a trivial thing. Thank you @AndreasHeine
